client/beefy/src/communication/request_response/outgoing_requests_engine.rs

```diff
@@ -0,0 +1,246 @@
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Generating request logic for request/response protocol for syncing BEEFY justifications.
+
+use beefy_primitives::{crypto::AuthorityId, BeefyApi, ValidatorSet};
+use codec::Encode;
+use futures::{
+	channel::{oneshot, oneshot::Canceled},
+	stream::{self, StreamExt},
+};
+use log::{debug, error, warn};
+use parking_lot::Mutex;
+use sc_network::{PeerId, ProtocolName};
+use sc_network_common::{
+	request_responses::{IfDisconnected, RequestFailure},
+	service::NetworkRequest,
+};
+use sp_api::ProvideRuntimeApi;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block, NumberFor},
+};
+use std::{collections::VecDeque, result::Result, sync::Arc};
+
+use crate::{
+	communication::request_response::{Error, JustificationRequest},
+	justification::{decode_and_verify_finality_proof, BeefyVersionedFinalityProof},
+	KnownPeers,
+};
+
+/// Response type received from network.
+type Response = Result<Vec<u8>, RequestFailure>;
+/// Used to receive a response from the network.
+type ResponseReceiver = oneshot::Receiver<Response>;
+
+enum State<B: Block> {
+	Idle(stream::Pending<Result<Response, Canceled>>),
+	AwaitingResponse(PeerId, NumberFor<B>, stream::Once<ResponseReceiver>),
+}
+
+pub struct OnDemandJustificationsEngine<B: Block, N, R> {
+	network: N,
+	runtime: Arc<R>,
+	protocol_name: ProtocolName,
+
+	live_peers: Arc<Mutex<KnownPeers<B>>>,
+	peers_cache: VecDeque<PeerId>,
+
+	state: State<B>,
+}
+
+impl<B, N, R> OnDemandJustificationsEngine<B, N, R>
+where
+	B: Block,
+	N: NetworkRequest,
+	R: ProvideRuntimeApi<B>,
+	R::Api: BeefyApi<B>,
+{
+	pub fn new(
+		network: N,
+		runtime: Arc<R>,
+		protocol_name: ProtocolName,
+		live_peers: Arc<Mutex<KnownPeers<B>>>,
+	) -> Self {
+		Self {
+			network,
+			runtime,
+			protocol_name,
+			live_peers,
+			peers_cache: VecDeque::new(),
+			state: State::Idle(stream::pending()),
+		}
+	}
+
+	fn reset_peers_cache_for_block(&mut self, block: NumberFor<B>) {
+		// TODO (issue #12296): replace peer selection with generic one that involves all protocols.
+		self.peers_cache = self.live_peers.lock().at_least_at_block(block);
+	}
+
+	fn try_next_peer(&mut self) -> Option<PeerId> {
+		// TODO (issue #12296): replace peer selection with generic one that involves all protocols.
+		let live = self.live_peers.lock();
+		while let Some(peer) = self.peers_cache.pop_front() {
+			if live.contains(&peer) {
+				return Some(peer)
+			}
+		}
+		None
+	}
+
+	fn request_from_peer(&mut self, peer: PeerId, block: NumberFor<B>) {
+		debug!(target: "beefy::sync", "🥩 requesting justif #{:?} from peer {:?}", block, peer);
+
+		let payload = JustificationRequest::<B> { begin: block }.encode();
+
+		let (tx, rx) = oneshot::channel();
+
+		self.network.start_request(
+			peer,
+			self.protocol_name.clone(),
+			payload,
+			tx,
+			IfDisconnected::ImmediateError,
+		);
+
+		self.state = State::AwaitingResponse(peer, block, stream::once(rx));
```

Comment:
Out of curiosity, why did you choose to wrap the `oneshot::Receiver` inside `stream::Once`?

Reply:
Yeah, I actually struggled with this part for a good while.

To cleanly add `OnDemandJustificationsEngine` to the worker's main futures driving loop we wanted this `Stream`-like behavior:
`OnDemandJustificationsEngine::next()` is called inside a loop under a `futures::select!` block, meaning each branch there is tried/polled at each iteration of the loop.

Our inner `OnDemandJustificationsEngine::next()` implementation is basically a future wrapped around `oneshot::Receiver` (`rx` in this line here), **but** the problem with oneshot is that it consumes itself when `await`ing, so on the first call to `next()` the receiver is consumed to create a `Future` object even if that future is not ready yet, then the `futures::select` invisibly drops this returned future when it selects some other (ready) one to return.

Initially I had this working by adding (wrapping) `oneshot::Receiver` to a `FuturesUnordered` as that doesn't consume itself when `await`ing. On a subsequent pass I switched it to using this simpler `Stream` model.

With `stream::once()` we get a stream that we can call `next()` on repeatedly, but it will only actually move forward when one of those returned futures is executed/completed, and intermediate ones can safely be dropped.

Reply:
Thank you for the explanation. Very interesting and an elegant solution as well I think.

Reply:
The solution to this would have been to use `(&mut receiver).await`.